packages/contracts/src/contracts/current/forwarder/MixinBuyExactTokens.sol

```diff
@@ -0,0 +1,128 @@
+pragma solidity ^0.4.23;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+import "./MixinForwarderExpectedResults.sol";
+import "./MixinERC20.sol";
+import "./MixinERC721.sol";
+
+contract MixinBuyExactTokens is
+    MixinForwarderCore,
+    MixinForwarderExpectedResults,
+    MixinERC20,
+    MixinERC721
+{
+    /// @dev Buys the exact amount of tokens (ERC20 and ERC721), performing fee abstraction if required. Deducts a proportional fee to fee recipient.
+    ///      This function is payable and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      The caller is sent all tokens from the fill of orders. This function will revert unless the requested amount of tokens are purchased.
+    ///      Any excess ETH sent will be returned to the caller
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param tokenAmount The amount of maker asset tokens to buy. Optional for ERC721.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the incoming ETH and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return FillResults amounts filled and fees paid by maker and taker.
+    function buyExactTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 tokenAmount,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        require(msg.value > 0, "msg.value must be greater than 0");
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), "order taker asset must be Wrapped ETH");
+        uint8 proxyId = uint8(orders[0].makerAssetData[0]);
+        require(proxyId == 1 || proxyId == 2, "unsupported token proxy");
+
+        uint256 remainingTakerTokenAmount = msg.value;
+        ETHER_TOKEN.deposit.value(remainingTakerTokenAmount)();
+        if (proxyId == 1) {
```

Comment:
Yeah the calculation is different for ERC721 (we sum all fees) versus ERC20 (we calculate expected results) and the scenario for buying ERC20 ZRX is different (we buy from the orders rather than feeOrders as its cheaper than 2 fills)